### ISSUE TITLE
Revisit panicking in `std::fmt::Display` implementations

### ISSUE THREAD

**Issue by Veetaha** [rust-itertools/itertools#667]
Panicking in `std::fmt::Display` implementations is a glaring footgun. Here is an example, where you are guaranteed to spend several hours debugging it, and potentially you will never find a cause because the panic happens in production deeply inside the logging system, which is your eyes, but they are now blind:

```
use std::collections::HashMap;
use tracing_subscriber::prelude::*;
use itertools::Itertools;


let (loki, _join_handle) = tracing_loki::layer(
    "http://loki:3100".parse().unwrap(),
    HashMap::new(),
    HashMap::new(),
)
.unwrap();

tracing_subscriber::registry()
    .with(tracing_subscriber::fmt::layer())
    // Commenting out the following line fixes the panic, but it's not obvious.
    // Basically any other layer that runs the `fmt::Display` implementation
    // of the `Itertools::format` second time after `fmt` subscriber may suffer from this panic
    .with(loki) 
    .init();
    
std::panic::set_hook(Box::new(move |panic_info| {
    tracing::error!("panic happened");
}));    

// panics here deep inside tracing when several layers display the value
tracing::info!(field = %[1, 2, 3].iter().format(","), "Log event"); 
```

With this you'll see only this log output without any indication of panic:

<img width="812" alt="image" src="https://user-images.githubusercontent.com/36276403/209455497-f1fcad83-17c6-4f1b-8380-f3b49a499011.png">

# Solution

Don't panic in `std::fmt::Display` implementations. Instead, the best approach would be to format a message with a lot of CAPS letters that say what the problem there is.

**Comment by Philippe-Cholet** [comment]
This definitely is a footgun. We panic in two `Display` implementations (only): `Format` (#307) and `FormatWith`.
#307 mentions the possibility of adding a `Clone` bound but that would be a nasty breaking change.

@phimuemue I would be in favor of formatting an error message rather than panicking.

**Comment by phimuemue** [comment]
I agree it is concerning that our `Display` implementation can panic. But is an error message upon the second try *really* better? I would find an error message for `Debug` okish ([its documentation](https://doc.rust-lang.org/std/fmt/trait.Debug.html) hints at the fact that `Debug`s output is not stable), not so much for `Display` ([which](https://doc.rust-lang.org/std/fmt/trait.Display.html) is for user-facing output):

I think that people rely on the output of `Display` (I certainly do) and are better off if they know when there's something strange instead of having an arbitrary error message (that might even change over time) that causes problems down the line. In my book, returning the error message is simply incorrect for `Display`.

So, we could rename it to `lazy_but_possibly_panicking_format`, or deprecate it and only keep something that converts to `String` once.

__Possibly off-topic:__

I am unsure, but may the general problem the following: [`Display::fmt`](https://doc.rust-lang.org/std/fmt/trait.Display.html#tymethod.fmt) takes `&self`, but there apparently are things that probably need `self` to be displayed. If `Display` was actually this:

```
trait HypotheticalDisplay {
 fn fmt(self, f: &mut Formatter) -> Result<(), Error>;
}
```

Then, you could - theoretically, and possibly using specialization - `impl<'a> HyoptheticalDisplay for &'a DisplayableByRef` vs `impl HypotheticalDisplay for OurIteratorFormatType`. Unsure if this would be a good idea, and I don't remember having seen it anywhere.

**Comment by Philippe-Cholet** [comment]
I see your point on incorrectness for `Display`.

Would it help _enough_ to panic with a big CAPS error message then? For the user to easily find it in a possibly large log?

    RUNTIME ERROR: `itertools::format` MUST BE FORMATTED ONLY ONCE, THE DEVELOPPER BROKE THE SPACE-TIME CONTINUUM

`lazy_but_possibly_panicking_format` would be quite long and alarming. Rename `format_once` would make the intent clearer but would not necessary help people as it might not be directly clear that the logger format the damn thing more than once.

We are not using const-generics yet so use specialization seems like an eternity to me at the moment. But it would be nicer!

**Comment by Veetaha** [comment]
Well, we have to make a judgement call here. We have to admit there isn't a perfect solution for this problem. We just have to pick our poison. I already consider `format` discouraged. Instead, I'm using `join(sep) -> String` instead to be safe. I don't think `format` should be marked as deprecated. However, I'd rather put big warning (⚠️ ⚠️ ⚠️) emojis at the top of its documentation saying that the method is logically panic-unsafe, and describe the potential footgun of its usage in `tracing` macros, which may format the same log field twice. In the example that I provided a mere addition of new `tracing` layer makes the object passed to a log macro formatted twice. I think when people see this in the docs they'll be warned enough and use `join()` instead to be safe if their code isn't performance-sensitive.

**Comment by Philippe-Cholet** [comment]
So merely update the docs, easy enough.
No CAPS error message?

**Comment by Veetaha** [comment]
> No CAPS error message?

If you mean formatting a CAPS message as output of display instead of panicking, that would also be cool, but I also see the point in producing a hard panic that would be much more visible than that (unless you use tracing in panic reporting), and I'm sure there will be issues complaining that they missed the message. If I were to re-do format from day zero, I'd avoid a panic, but today people probably expect it to panic already, so let's keep it this way

**Comment by Philippe-Cholet** [comment]
Oops, I meant panicking with an error message in CAPS like ``"`itertools::Format` WAS ALREADY FORMATTED ONCE"``. Would it help debugging if it comes to it?

**Comment by Veetaha** [comment]
Caps letters would be more visible indeed, but it wouldn't help in the case that I reported in my issue where the panic message was totally not visible because `tracing` was used in the panic handler. Anyway, I think panic itself is visible enough when its message is printed at all. Making it CAPS would probably be excessive as for me
